add mpls auto mode to update mpls control upon routing events

pguibert6WIND opened the pull request:
The yang NB API does not handle the mpls configuration on its leaf.
Add an mpls leaf to stick to the mpls configuration.
- true or false to mean if config
- not defined, means no config.

pguibert6WIND committed:
ldpd, lib, zebra, topotests: add MPLS interface config from LDP

When LDP is enabled per interface, MPLS is not turned on on each
of those interfaces.

Fix this by letting LDP configure the interface if the 'mpls auto'
mode is configured.

Create an IMSG to send interface config request from ldpe thread
to main thread of ldp.
Create a ZAPI message to set MPLS from ldp to zebra.
Update the MPLS setting.
Add a test that ensures that LDP properly configures the interface.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

mjstapp commented on zebra/zebra_nb_config.c:
```diff
@@ -1087,6 +1087,50 @@ int lib_interface_zebra_shutdown_destroy(struct nb_cb_destroy_args *args)
+int lib_interface_zebra_mpls_destroy(struct nb_cb_destroy_args *args)
+{
+	struct interface *ifp;
+	struct zebra_if *if_data;
```
the convention is to call these pointers "zif"

mjstapp commented on zebra/zebra_nb_config.c:
```diff
@@ -1087,6 +1087,50 @@ int lib_interface_zebra_shutdown_destroy(struct nb_cb_destroy_args *args)
+
+	if_data->mpls_config = IF_ZEBRA_DATA_UNSPEC;
+
+	dplane_intf_mpls_modify_state(ifp, false);
```
so ... this is sort of a problem we've had in this area. a 'destroy' should mean 'no configuration', but this looks like "configured to 'off'".

pguibert6WIND replied:
you are right. I added a commit to fix this issue.

mjstapp commented:
I'm a little uncertain what this nb work means. Is this actually plumbed through zebra - many of the apis look to be stubs? There's no mgmtd connection, right? so ... who's talking to zebra to effect this mpls config change via the nb?

pguibert6WIND committed:
zebra, yang, doc: add 'mpls auto' command

There is not auto mode for mpls settings, so that
when any labeled route or mpls entry is configured,
then the used interface should automatically turn
on its mpls settings.

Add the following per interface command:

> mpls auto

Change the yang model to replace the mpls boolean
value with an enumerate, which is mapped over the
data values declared under zebra/interface.h.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
zebra: add dynamic mpls activation upon routing events

When a labeled route, or an MPLS entry is configured, the
interface may not have properly enabled the interface
with mpls.

When the interface is in mpls auto mode, change the mpls
interface value when a route or an mpls entry requests
it to be turned on.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND commented:
> I'm a little uncertain what this nb work means. Is this actually plumbed through zebra - many of the apis look to be stubs? There's no mgmtd connection, right? so ... who's talking to zebra to effect this mpls config change via the nb?

The NB API is supposed to completely replace the 'struct interface' support, when it deals with configuration. So I guess this is where we have to go to, even if the path is not completely clear to me.
What I can say for instance is that this patch https://github.com/FRRouting/frr/commit/26f8f6fe7fb9 should one day in the future removed, and we will only use the built yang tree to recover interfaces.

pguibert6WIND commented:
ci:rerun

pguibert6WIND commented:
ci:rerun

ton31337 commented on zebra/interface.h:
```diff
@@ -26,6 +26,7 @@ extern "C" {
 #define IF_ZEBRA_DATA_UNSPEC 0
 #define IF_ZEBRA_DATA_ON 1
 #define IF_ZEBRA_DATA_OFF 2
+#define IF_ZEBRA_DATA_AUTO   3
```
Can we use the same spacing as above?

pguibert6WIND replied:
done

ton31337 reviewed (changes requested):
Documentation is missing.

pguibert6WIND commented on zebra/interface.c:
```diff
@@ -3775,25 +3778,19 @@ DEFUN (multicast,
-       NO_STR
-       MPLS_STR
-       "Set mpls to be on for the interface\n")
+DEFPY(mpls, mpls_cmd, "[no] mpls <enable|disable|auto>$state",
```
done

pguibert6WIND commented:
> Documentation is missing.

done

mjstapp commented:
whoa, whoa - the description here just talks about making a northbound/yang-model change, but there's a number of new pieces of dynamic logic that appears to be switching mpls forwarding on and off during rib and lsp processing. that needs to be, at a bare minimum, described clearly, and preferably be in a separate PR, separate from the mechanical yang-model changes.

pguibert6WIND commented:
> whoa, whoa - the description here just talks about making a northbound/yang-model change, but there's a number of new pieces of dynamic logic that appears to be switching mpls forwarding on and off during rib and lsp processing. that needs to be, at a bare minimum, described clearly, and preferably be in a separate PR, separate from the mechanical yang-model changes.

You are right to mention two changes in a single PR. however, the auto mode does not take too much code. and it is a separate commit.
to get more clarity about https://github.com/FRRouting/frr/pull/13965/commits/b53fccfdfdaf59e52ed25116dbeef8c556bc738e, I picked up routing events and upgraded the mpls control every time needed: route add, mpls add events.

pguibert6WIND committed:
topotests: add static mpls routing auto mode

There is no test that checks for the mpls auto mode
configuration.
New checks are added to the static mpls routing test:
- check that a labeled route turns on mpls on the interface
where the auto mode is set
- check that a configured interface recovers the mpls setting

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND commented:
ci:rerun

pguibert6WIND commented:
@mjstapp , for MPLS Yang model and separate PR, please see https://github.com/FRRouting/frr/pull/14176.

louis-6wind commented:
ci:rerun

pguibert6WIND committed:
zebra: add default MPLS interface settings framework

By default, MPLS configuration per interface is unspecified,
but can not be changed by compilation.
Add the framework to change the MPLS settings.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
zebra: add json support for mpls config per interface

Add 'mplsConfig' json option, that defines which MPLS
configuration is applied to a given interface.

> # show interface loop1 json
> {
>   "loop1":{
> [..]
>   "mplsConfig":"mplsEnabled",
>   "mplsEnabled":false,
> [..]

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
zebra: change default mpls interface settings to auto mode

Change the default MPLS setting per interface to auto mode.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
lib: add default support for uint8_t variable

To support the various values of mpls configuration,
as the value is encoded in a 8 bits value, define
a new uint8_t type to record it.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND commented:
ci:rerun

ton31337 commented on lib/ldp_sync.h:
```diff
@@ -56,6 +56,11 @@ struct ldp_igp_sync_if_state {
 
+struct ldp_igp_configure_if_mpls {
+	ifindex_t ifindex;
+	bool status;
```
if the status is enabled/disabled, then it's more clear it would be named `bool enabled` (IMO)... Because status could be more than enabled/disabled, right?

pguibert6WIND replied:
changed to enabled.

ton31337 commented on lib/zclient.c:
```diff
@@ -2864,6 +2864,52 @@ size_t zebra_interface_link_params_write(struct stream *s,
+				  ifindex_t ifindex, bool set)
+{
+	struct stream *s;
+	int w;
```
Where do you use this except for +?

pguibert6WIND replied:
fixed. removed the w

ton31337 commented on tests/topotests/static_routing_mpls/test_static_routing_mpls.py:
```diff
@@ -134,6 +152,89 @@ def test_mpls_configured_on_interface():
+        pytest.skip(tgen.errors)
+
+    router = tgen.gears["r2"]
+    output = router.vtysh_cmd("config terminal\nmpls lsp 33 192.168.2.3 implicit-null")
```
you don't check/print the output of `output`, we can just remove this assignment...

pguibert6WIND replied:
fixed

ton31337 commented on tests/topotests/static_routing_mpls/test_static_routing_mpls.py:
```diff
@@ -134,6 +152,89 @@ def test_mpls_configured_on_interface():
+        "r1, configuring a route with labeled nexthop address, checking that MPLS state is on on r1-eth1"
+    )
+    router = tgen.gears["r1"]
+    output = router.vtysh_cmd(
```
ditto

ton31337 commented on tests/topotests/static_routing_mpls/test_static_routing_mpls.py:
```diff
@@ -134,6 +152,89 @@ def test_mpls_configured_on_interface():
+        "r2, configuring a route with labeled nexthop interface, checking that MPLS state is on on r2-eth0"
+    )
+    router = tgen.gears["r2"]
+    output = router.vtysh_cmd(
```
ditto

pguibert6WIND replied:
fixed

pguibert6WIND replied:
fixed

ton31337 commented on tests/topotests/static_routing_mpls/test_static_routing_mpls.py:
```diff
@@ -134,6 +152,89 @@ def test_mpls_configured_on_interface():
+        "r1, configuring a route with labeled nexthop address, checking that MPLS is configured on r1-eth1.100"
+    )
+    router = tgen.gears["r1"]
+    output = router.vtysh_cmd(
```
ditto

pguibert6WIND replied:
fixed

ton31337 commented on zebra/interface.c:
```diff
@@ -49,6 +49,10 @@ DEFINE_HOOK(zebra_if_config_wr, (struct vty * vty, struct interface *ifp),
 DEFINE_MTYPE(ZEBRA, ZIF_DESC, "Intf desc");
 
+FRR_CFG_DEFAULT_UINT8_T(ZEBRA_MPLS,
+	{ .val_uint8_t = IF_ZEBRA_DATA_AUTO},
```
styling: missing whitespace after AUTO:
```c
{ .val_uint8_t = IF_ZEBRA_DATA_AUTO },
```

pguibert6WIND replied:
fixed.

ton31337 commented on zebra/interface.c:
```diff
@@ -5344,6 +5367,59 @@ static int ipv6_address_install(struct vty *vty, struct interface *ifp,
+
+	if (!ifp)
+		return;
+	zif = ifp->info;
```
Shouldn't we check if zif is not NULL?

pguibert6WIND replied:
yes.done

ton31337 commented on zebra/interface.h:
```diff
@@ -116,6 +118,9 @@ struct zebra_if {
 	uint8_t mpls_config;
 
+	/* MPLS dynamic configuration - when mpls_config is auto */
+	uint8_t mpls_dynamic;
```
What values can it have actually?

ton31337 commented on zebra/zapi_msg.c:
```diff
@@ -1791,7 +1792,9 @@ static bool zapi_read_nexthops(struct zserv *client, struct prefix *p,
 		    && api_nh->type != NEXTHOP_TYPE_IFINDEX
 		    && api_nh->type != NEXTHOP_TYPE_BLACKHOLE
 		    && api_nh->label_num > 0) {
-
```
Please drop this empty line

pguibert6WIND replied:
done

ton31337 commented on zebra/zapi_msg.c:
```diff
@@ -3182,16 +3185,19 @@ static void zread_vrf_label(ZAPI_HANDLER_ARGS)
 		mpls_label_t out_label = MPLS_LABEL_IMPLICIT_NULL;
 		mpls_lsp_install(def_zvrf, ltype, nlabel, 1, &out_label,
 				 NEXTHOP_TYPE_IFINDEX, NULL, ifp->ifindex);
+		mpls_auto_interface_data_on(ifp);
```
Shouldn't this be turned on before installing MPLS LSP?

pguibert6WIND replied:
no, this should be turned on, only when used.
actually, setting the /proc/sys/net/mpls/input flag may have negative impact on kernel, when interface flapping happens, and when platform_auto value has huge value. 
this method avoids generalising the usage of the input flag on every interface.

louis-6wind commented on zebra/zebra_cli.c:
```diff
@@ -11,14 +11,11 @@
-	{ .val_bool = true, .match_profile = "traditional", },
-	{ .val_bool = false },
-);
-
```
this is unrelated

louis-6wind commented on zebra/zebra_vrf.h:
```diff
@@ -19,11 +19,6 @@
-	{ .val_bool = true, .match_profile = "traditional", },
-	{ .val_bool = false },
-);
-
```
this is unrelated

riw777 commented:
@frrbot autoclose in 2 weeks
